pyproject.toml

```diff
@@ -97,10 +97,10 @@ maintenance = ["conda-lock==2.5.7"]
 build-backend = "mesonpy"
 # Minimum requirements for the build system to execute.
 requires = [
-    "meson-python>=0.16.0",
-    "Cython>=3.0.10",
-    "numpy>=2",
-    "scipy>=1.8.0",
+    "meson-python>=0.16.0,<0.19.0",
+    "Cython>=3.0.10,<3.0.13",
+    "numpy>=2,<2.3.0",
+    "scipy>=1.8.0,<1.16.0",
 ]
```

Comment:
upper bound the build dependencies in release branch. See discussions in #31183

I used the same policy as scipy, i.e upper bound to the last major/minor release at the time (allow future bug-fix releases).

cc/ @thomasjpfan @ogrisel @glemaitre @lesteve @betatim @adrinjalali

Reply:
I'm okay with using the SciPy approach. (We'll need to add this policy to the release checklist)

Given this is a build dependency in `pyproject.toml`, distro packages can still choose to ignore them.

Reply:
I fixed the cython upper bound in [9f3dd26](https://github.com/scikit-learn/scikit-learn/pull/31335/commits/9f3dd26d833e599f1372ecbad94d09c650a504a3), because I previously used a x.y.z version, not following the policy I just mentioned...

>  (We'll need to add this policy to the release checklist)

I'll make a PR to main when the rc is out.

Reply:
So I had to modify the cython upper bound again because we need a pre-release for free-threading for now.
The current official releease is ``3.0.12`` so the upper bound should be ``3.1.0``.
But free-threading support is only available in ``3.1.Zrc1``, so I set the upper bound to ``3.2.0``.

Reply:
Since the nightly builds are now green with 3.1.0rc1, I think it's fine.